config/src/keys.rs

```diff
@@ -0,0 +1,233 @@
+    #[serde(deserialize_with = "deserialize_key")]
+    network_signing_public_key: Ed25519PublicKey,
+
+    #[serde(serialize_with = "serialize_key")]
+    #[serde(deserialize_with = "deserialize_key")]
+    network_identity_private_key: X25519StaticPrivateKey,
+    #[serde(serialize_with = "serialize_key")]
+    #[serde(deserialize_with = "deserialize_key")]
+    network_identity_public_key: X25519StaticPublicKey,
+}
+
+// required for serialization
+impl Default for NetworkKeyPairs {
+    fn default() -> Self {
+        let mut rng = StdRng::from_seed(TEST_SEED);
+        let (net_private_sig, net_public_sig) = compat::generate_keypair(&mut rng);
+        let (private_kex, public_kex) = x25519::compat::generate_keypair(&mut rng);
+        Self {
+            network_signing_private_key: PrivateKeyContainer::Present(net_private_sig),
+            network_signing_public_key: net_public_sig,
+            network_identity_private_key: private_kex,
+            network_identity_public_key: public_kex,
+        }
+    }
+}
+
+impl NetworkKeyPairs {
+    // used to deserialize keypairs from a configuration file
+    pub fn load_config<P: AsRef<Path>>(path: P) -> Self {
+        let path = path.as_ref();
+        let mut file = File::open(path)
+            .unwrap_or_else(|_| panic!("Cannot open Network keypair Config file {:?}", path));
+        let mut contents = String::new();
+        file.read_to_string(&mut contents)
+            .unwrap_or_else(|_| panic!("Error reading Network keypair Config file {:?}", path));
+        Self::parse(&contents)
+    }
+
+    fn parse(config_string: &str) -> Self {
+        toml::from_str(config_string).expect("Unable to parse Config")
+    }
+    // used to serialize keypairs to a configuration file
+    pub fn save_config<P: AsRef<Path>>(&self, output_file: P) {
+        let contents = toml::to_vec(&self).expect("Error serializing");
+        let mut file = File::create(output_file).expect("Error opening file");
+        file.write_all(&contents).expect("Error writing file");
+    }
+
+    // used in testing to fill the structure with test keypairs
+    pub fn load(
+        network_signing_private_key: Ed25519PrivateKey,
+        network_identity_private_key: X25519StaticPrivateKey,
+    ) -> Self {
+        let network_signing_public_key = (&network_signing_private_key).into();
+        let network_identity_public_key = (&network_identity_private_key).into();
+        Self {
+            network_signing_private_key: PrivateKeyContainer::Present(network_signing_private_key),
+            network_signing_public_key,
+            network_identity_private_key,
+            network_identity_public_key,
+        }
+    }
+
+    /// Beware, this destroys the private key from this NodeConfig
+    pub fn take_network_signing_private(&mut self) -> Option<Ed25519PrivateKey> {
+        self.network_signing_private_key.take()
+    }
+
+    pub fn get_network_identity_private(&self) -> X25519StaticPrivateKey {
```

Comment:
does this need a testing feature gate, since it clones a private key?

Reply:
yes